### DISCUSSION TITLE
Section 14: Understanding SmallProxy Contract Behavior with ImplementationA

### DISCUSSION THREAD

**Discussion by vascofcd** [Cyfrin/foundry-full-course-cu#2029] [Q&A]
Hello everyone,

I'm currently working through the lesson 14 about Foundry Upgrades and decided to do some experiments with the `SmallProxy` contract. Below is the code for the `SmallProxy` and `ImplementationA` contracts (it was modified!):

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "@openzeppelin/contracts/proxy/Proxy.sol";

contract SmallProxy is Proxy {
    address public admin; //added this

    bytes32 private constant _IMPLEMENTATION_SLOT = 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;

    //added this
    function setSomeAddress(address newAdmin) external {
        admin = newAdmin;
    }

    function setImplementation(address newImplementation) public {
        assembly {
            sstore(_IMPLEMENTATION_SLOT, newImplementation)
        }
    }

    function _implementation() internal view override returns (address implementationAddress) {
        assembly {
            implementationAddress := sload(_IMPLEMENTATION_SLOT)
        }
    }

    // helper function
    function getDataToTransact(uint256 numberToUpdate) public pure returns (bytes memory) {
        return abi.encodeWithSignature("setValue(uint256)", numberToUpdate);
    }

    function readStorage() public view returns (uint256 valueAtStorageSlotZero) {
        assembly {
            valueAtStorageSlotZero := sload(0)
        }
    }
}

contract ImplementationA {
    address public owner;

    function setOwner() public {
        owner = msg.sender;
    }
}
```
Here is the unit test I wrote for these contracts:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {Test, console} from "forge-std/Test.sol";
import {ImplementationA, SmallProxy} from "../../src/24-PuzzleWallet/ProxyExample/SmallProxy.sol";

contract SmallProxyTest is Test {
    ImplementationA implementationA;
    SmallProxy smallProxy;

    address owner = makeAddr("owner");
    address attacker = makeAddr("attacker");

    function setUp() public {
        vm.startPrank(owner, owner);

        // init impl contract
        implementationA = new ImplementationA();

        // setup proxy
        smallProxy = new SmallProxy();
        smallProxy.setImplementation(address(implementationA));
        implementationA = ImplementationA(address(smallProxy));
        
        vm.stopPrank();
    
        assertEq(address(0), smallProxy.admin()); // passed
        assertEq(address(0), implementationA.owner()); // passed
    }

    function testStorageLayout() public {
        // part 1 owner to attacker
        vm.startPrank(attacker, attacker);
        smallProxy.setSomeAddress(attacker);
        vm.stopPrank();

        assertEq(attacker, smallProxy.admin()); // passed
        assertEq(attacker, implementationA.owner()); // passed

        // part 2 attacker to owner
        vm.startPrank(owner, owner);
        implementationA.setOwner();
        vm.stopPrank();

        assertEq(owner, smallProxy.admin()); // passed
        assertEq(owner, implementationA.owner()); // passed
    }
}
```
First, I'm assuming that `ImplementationA` and `SmallProxyContract` are two independent contracts, each one with its own storage layout. I'm puzzled by the behavior of:

## Changing the Admin Address

- `SmallProxy::setSomeAddress` is supposed to set a new address for the admin variable in slot 0 of `SmallProxy`.
- However, `implementationA.owner` is also changed to the attacker’s address, despite no `delegatecall` being present in `SmallProxy::setSomeAddress`.
- I expected only `SmallProxy.admin` to change, but both are being updated. Why is this happening?

## Changing the Owner Address

- When calling `implementationA.setOwner()`, both `admin` and `owner` are updated to the owner's address.
- Again, there’s no `delegatecall` in this function, so why are both variables being updated?

Now this contract utilizes `Proxy` from OpenZeppelin, so my guess is that something is happening behind the scenes with the `fallback` and `delegatecall` function that I'm not able to understand.

2. Can someone provide a mental model to understand this behavior? What is really happening step by step? 
3. Are there any resources or documentation that could help clarify this?

Thank you in advance!

**Comment by EngrPips** [comment]
Awesome. this will be an interesting one.

**Comment by caodanbobo** [comment]
I am also new to Solidity with java background. 
Based on my experience from java inheritance, the 'implementationA' would point to the instance of 'ImplementationA' after following command executed. 
```
implementationA = ImplementationA(address(smallProxy));
```
However, it acts differently from Java. In Solidity, casting an address to a contract type is essentially telling the compiler that you expect the contract at this address to conform to the interface of the specified type. 
In other words, the 'implementationA' is still a 'smallProxy', because they share the same address.
I hope this will help.

**Reply by caodanbobo** [reply]
i was talking about 'inheritance' because i was confused how inheritance works in solidlity.
FYI: https://github.com/Cyfrin/foundry-full-course-cu/discussions/2013#discussioncomment-10046674
in the context of your question, there is nothing to do with inheritance, and i am not sure what was 'implementationA = ImplementationA(address(smallProxy));' for.

**Reply by EngrPips** [reply]
This `implementationA = ImplementationA(address(smallProxy));` simply means we wrap the address of the proxy contract with the Interface of the `ImplementationA` contract so that we can directly call the `ImplementationA` contract methods on the proxy contract, which the proxy contract will further delegate to the implementation contract.
